module/apmfasthttp/utils.go

```diff
@@ -0,0 +1,114 @@
+package apmfasthttp
+
+import (
+	"net/http"
+	"net/url"
+
+	"github.com/savsgio/gotils/strconv"
+	"github.com/valyala/bytebufferpool"
+	"github.com/valyala/fasthttp"
+	"go.elastic.co/apm"
+	"go.elastic.co/apm/module/apmhttp"
+)
+
+func resetHTTPRequest(req *http.Request) {
+	req.Method = ""
+	req.URL = nil
+	req.Proto = ""
+	req.ProtoMajor = 0
+	req.ProtoMinor = 0
+	resetHTTPMap(req.Header)
+	req.Body = nil
+	req.GetBody = nil
+	req.ContentLength = 0
+	req.TransferEncoding = req.TransferEncoding[:0]
+	req.Close = false
+	req.Host = ""
+	resetHTTPMap(req.Form)
+	resetHTTPMap(req.PostForm)
+	req.MultipartForm = nil
+	resetHTTPMap(req.Trailer)
+	req.RemoteAddr = ""
+	req.RequestURI = ""
+	req.TLS = nil
+	// req.Cancel = nil
+	req.Response = nil
+}
+
+func resetHTTPMap(m map[string][]string) {
+	for k := range m {
+		delete(m, k)
+	}
+}
+
+func requestCtxToRequest(ctx *fasthttp.RequestCtx, req *http.Request, httpBody *netHTTPBody) error {
+	body := ctx.Request.Body()
+
+	req.Method = strconv.B2S(ctx.Method())
+	req.Proto = "HTTP/1.1"
+	req.ProtoMajor = 1
+	req.ProtoMinor = 1
+	req.RequestURI = strconv.B2S(ctx.RequestURI())
+	req.ContentLength = int64(len(body))
+	req.Host = strconv.B2S(ctx.Host())
+	req.RemoteAddr = ctx.RemoteAddr().String()
+	req.TLS = ctx.TLSConnectionState()
+
+	httpBody.b = append(httpBody.b[:0], body...)
+	req.Body = httpBody
```

Comment:
To know the `req.ContenLenght` the body is preparsed into a `[]body` (independently of streaming or not), so here in both cases, the `ctx.RequestBodyStream` is not so helpful, i think

Reply:
And the `ctx.RequestBodyStream` just works if  (fasthttp.Server.StreamRequestBody)[https://github.com/valyala/fasthttp/blob/master/server.go#L389] is enabled.

Reply:
I took a look at the implementation of `Request.Body` and it appears to consume the entire request body stream: https://github.com/valyala/fasthttp/blob/620f0c83adb3d72021545ecbb47ff1bef42fbd0c/http.go#L351-L359

That means the middleware will block until the request body stream is completed, which may alter application behaviour. Furthermore, the request body stream could be very large and this will read it all into memory.

I think the safe thing to do here is:

 - check Request.IsBodyStream; if true, set `req.Body` to `ioutil.NopCloser(ctx.RequestBodyStream())`
 - otherwise set `req.Body` to `ioutil.NopCloser(bytes.NewReader(ctx.Request.Body()))`